compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -544,6 +554,53 @@ impl<'a> Parser<'a> {
         Ok(TyKind::BareFn(P(BareFnTy { ext, unsafety, generic_params: params, decl, decl_span })))
     }
 
+    /// Recover from function pointer types with a generic parameter list (e.g. `fn<'a>(&'a str)`).
+    fn recover_fn_ptr_with_generics(
+        &mut self,
+        lo: Span,
+        params: &mut Vec<GenericParam>,
+        param_insertion_point: Option<Span>,
+    ) -> PResult<'a, ()> {
+        let generics = self.parse_generics()?;
+
+        let mut lifetimes: Vec<_> = generics
+            .params
+            .into_iter()
+            .filter(|param| matches!(param.kind, ast::GenericParamKind::Lifetime))
+            .collect();
+
+        let sugg = if !lifetimes.is_empty() {
+            let snippet = lifetimes
+                .iter()
+                .map(|param| param.ident.as_str())
+                .intersperse(", ")
+                .collect::<String>();
+
+            let (left, snippet) = if let Some(span) = param_insertion_point {
+                (span, if params.is_empty() { snippet } else { format!(", {snippet}") })
+            } else {
+                (lo.shrink_to_lo(), format!("for<{snippet}> "))
+            };
+
+            Some(FnPtrWithGenericsSugg {
+                left,
+                snippet,
+                right: generics.span,
+                for_param_list_exists: param_insertion_point.is_some(),
+            })
+        } else {
+            None
+        };
+
+        self.sess.emit_err(FnPtrWithGenerics { span: generics.span, sugg });
+
+        // Only pass along lifetime parameters since type & const parameters aren't
+        // allowed to appear in `for` parameter lists.
```

Comment:
I'd like to clarify this comment (I should probably reword it): I could easily pass along every kind of parameter (lt, ty, ct).

However, the subsequent error message would be *only lifetime parameters can be used in this context* pointing at the incorrectly positioned parameter list which clearly isn't great (in my eyes, it's confusing & contradictory). If the message was sth. akin to *$kind parameter cannot be used in this context* (similar to the *lifetime bounds cannot be used in this context* message), it would be better in this scenario (it's neutral). I am not suggesting we should change it though.